File: sprints/rag_engine/rag-engine-mini/sdk/python/rag_engine.py
# Python SDK for RAG Engine
# ==============================

"""
RAG Engine Python SDK
======================

Official Python SDK for RAG Engine API.

RAG Engine Python SDK الرسمي
"""

from typing import Optional, List, Dict, Any
from dataclasses import dataclass
import httpx
import asyncio
from enum import Enum


class DocumentStatus(str, Enum):
    """Document processing status."""

    CREATED = "created"
    INDEXED = "indexed"
    FAILED = "failed"


class QuerySortBy(str, Enum):
    """Query sorting options."""

    CREATED = "created"
    UPDATED = "updated"
    FILENAME = "filename"
    SIZE = "size"


@dataclass
class Document:
    """Document metadata."""

    id: str
    filename: str
    content_type: str
    size_bytes: int
    status: DocumentStatus
    created_at: str
    updated_at: Optional[str] = None


@dataclass
class Answer:
    """RAG answer with sources."""

    text: str
    sources: List[str]
    retrieval_k: int
    embed_ms: Optional[int] = None
    search_ms: Optional[int] = None
    llm_ms: Optional[int] = None


@dataclass
class QueryHistoryItem:
    """Query history item."""

    question: str
    answer: str
    sources: List[str]
    timestamp: str
    success: bool


class RAGClient:
    """
    RAG Engine API client.

    عميل API لـ RAG Engine
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.rag-engine.com",
        timeout: float = 30.0,
    ):
        """
        Initialize RAG Engine client.

        Args:
            api_key: API key for authentication
            base_url: Base URL of RAG Engine API
            timeout: Request timeout in seconds

        تهيئة عميل RAG Engine
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

        self._client: Optional[httpx.Client] = None
        self._async_client: Optional[httpx.AsyncClient] = None

    def _get_headers(self) -> Dict[str, str]:
        """Get request headers with authentication."""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    # -------------------------------------------------------------------------
    # Synchronous API
    # -------------------------------------------------------------------------

    def _get_sync_client(self) -> httpx.Client:
        """Get or create sync HTTP client."""
        if self._client is None:
            self._client = httpx.Client(
                base_url=self.base_url,
                timeout=self.timeout,
                headers=self._get_headers(),
            )
        return self._client

    def ask(
        self,
        question: str,
        k: int = 5,
        document_id: Optional[str] = None,
        expand_query: bool = False,
    ) -> Answer:
        """
        Ask a question to the RAG engine.

        Args:
            question: Question to ask
            k: Number of chunks to retrieve (default: 5)
            document_id: Optional document ID for chat mode
            expand_query: Use query expansion (default: False)

        Returns:
            Answer object with text and sources

        طرح سؤال على محرك RAG
        """
        client = self._get_sync_client()

        response = client.post(
            "/api/v1/ask",
            json={
                "question": question,
                "k": k,
                "document_id": document_id,
                "expand_query": expand_query,
            },
        )
        response.raise_for_status()

        data = response.json()
        return Answer(
            text=data["answer"],
            sources=data.get("sources", []),
            retrieval_k=data.get("retrieval_k", 0),
            embed_ms=data.get("embed_ms"),
            search_ms=data.get("search_ms"),
            llm_ms=data.get("llm_ms"),
        )

    def upload_document(
        self,
        file_path: str,
        filename: Optional[str] = None,
    ) -> Document:
        """
        Upload a document to the RAG engine.

        Args:
            file_path: Path to file to upload
            filename: Optional custom filename

        Returns:
            Document metadata

        رفع مستند إلى محرك RAG
        """
        client = self._get_sync_client()

        with open(file_path, "rb") as f:
            files = {
                "file": (filename or file_path.split("/")[-1], f),
            }
            response = client.post(
                "/api/v1/documents",
                files=files,
            )

        response.raise_for_status()

        data = response.json()
        return Document(
            id=data["id"],
            filename=data["filename"],
            content_type=data["content_type"],
            size_bytes=data["size_bytes"],
            status=DocumentStatus(data["status"]),
            created_at=data["created_at"],
        )

    def search_documents(
        self,
        query: str,
        k: int = 10,
        filters: Optional[Dict[str, Any]] = None,
        sort_by: QuerySortBy = QuerySortBy.CREATED,
        limit: int = 20,
        offset: int = 0,
    ) -> List[Document]:
        """
        Search for documents.

        Args:
            query: Search query
            k: Number of results (default: 10)
            filters: Optional filters (status, type, date, size)
            sort_by: Sort order (default: created)
            limit: Max results to return (default: 20)
            offset: Pagination offset (default: 0)

        Returns:
            List of matching documents

        البحث عن المستندات
        """
        client = self._get_sync_client()

        params = {
            "query": query,
            "k": k,
            "sort_by": sort_by.value,
            "limit": limit,
            "offset": offset,
        }

        if filters:
            params.update(filters)

        response = client.get(
            "/api/v1/documents/search",
            params=params,
        )
        response.raise_for_status()

        data = response.json()
        return [
            Document(
                id=d["id"],
                filename=d["filename"],
                content_type=d["content_type"],
                size_bytes=d["size_bytes"],
                status=DocumentStatus(d["status"]),
                created_at=d["created_at"],
                updated_at=d.get("updated_at"),
            )
            for d in data["results"]
        ]

    def delete_document(self, document_id: str) -> bool:
        """
        Delete a document.

        Args:
            document_id: ID of document to delete

        Returns:
            True if successful

        حذف مستند
        """
        client = self._get_sync_client()
        response = client.delete(f"/api/v1/documents/{document_id}")
        response.raise_for_status()
        return True

    def get_query_history(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> List[QueryHistoryItem]:
        """
        Get query history.

        Args:
            limit: Max history items (default: 50)
            offset: Pagination offset (default: 0)

        Returns:
            List of query history items

        الحصول على سجل الاستعلامات
        """
        client = self._get_sync_client()

        response = client.get(
            "/api/v1/queries/history",
            params={
                "limit": limit,
                "offset": offset,
            },
        )
        response.raise_for_status()

        data = response.json()
        return [
            QueryHistoryItem(
                question=q["question"],
                answer=q.get("answer", ""),
                sources=q.get("sources", []),
                timestamp=q["timestamp"],
                success=q.get("success", True),
            )
            for q in data["questions"]
        ]

    # -------------------------------------------------------------------------
    # Asynchronous API
    # -------------------------------------------------------------------------

    def _get_async_client(self) -> httpx.AsyncClient:
        """Get or create async HTTP client."""
        if self._async_client is None:
            self._async_client = httpx.AsyncClient(
                base_url=self.base_url,
                timeout=self.timeout,
                headers=self._get_headers(),
            )
        return self._async_client

    async def ask_async(
        self,
        question: str,
        k: int = 5,
        document_id: Optional[str] = None,
        expand_query: bool = False,
    ) -> Answer:
        """Async version of ask()."""
        client = self._get_async_client()

        response = await client.post(
            "/api/v1/ask",
            json={
                "question": question,
                "k": k,
                "document_id": document_id,
                "expand_query": expand_query,
            },
        )
        response.raise_for_status()

        data = response.json()
        return Answer(
            text=data["answer"],
            sources=data.get("sources", []),
            retrieval_k=data.get("retrieval_k", 0),
            embed_ms=data.get("embed_ms"),
            search_ms=data.get("search_ms"),
            llm_ms=data.get("llm_ms"),
        )

    async def upload_document_async(
        self,
        file_path: str,
        filename: Optional[str] = None,
    ) -> Document:
        """Async version of upload_document()."""
        client = self._get_async_client()

        with open(file_path, "rb") as f:
            files = {
                "file": (filename or file_path.split("/")[-1], f),
            }
            response = await client.post(
                "/api/v1/documents",
                files=files,
            )

        response.raise_for_status()

        data = response.json()
        return Document(
            id=data["id"],
            filename=data["filename"],
            content_type=data["content_type"],
            size_bytes=data["size_bytes"],
            status=DocumentStatus(data["status"]),
            created_at=data["created_at"],
        )

    def close(self):
        """Close HTTP clients and cleanup resources."""
        if self._client:
            self._client.close()
        if self._async_client:
            import asyncio

            asyncio.create_task(self._async_client.aclose())


# -----------------------------------------------------------------------------
# Example Usage
# -----------------------------------------------------------------------------

if __name__ == "__main__":
    # Initialize client
    client = RAGClient(
        api_key="sk_your_api_key_here",
        base_url="http://localhost:8000",
    )

    # Sync usage
    print("=== Sync Usage ===")
    answer = client.ask("What is RAG?", k=5)
    print(f"Answer: {answer.text}")
    print(f"Sources: {answer.sources}")

    # Async usage
    async def async_example():
        print("\n=== Async Usage ===")
        answer = await client.ask_async("How does vector search work?", k=10)
        print(f"Answer: {answer.text}")
        print(f"Sources: {answer.sources}")

    asyncio.run(async_example())

    # Cleanup
    client.close()

File: research/rag_engine/rag-engine-mini/sdk/javascript/rag-engine.js
# RAG Engine JavaScript SDK
# ==========================

/**
 * RAG Engine JavaScript/TypeScript SDK
 * =====================================
 * Official SDK for RAG Engine API (Node.js & Browser)
 *
 * RAG Engine JavaScript/TypeScript SDK الرسمي
 */

// Types
export enum DocumentStatus {
  CREATED = "created",
  INDEXED = "indexed",
  FAILED = "failed",
}

export enum QuerySortBy {
  CREATED = "created",
  UPDATED = "updated",
  FILENAME = "filename",
  SIZE = "size",
}

export interface Document {
  id: string;
  filename: string;
  contentType: string;
  sizeBytes: number;
  status: DocumentStatus;
  createdAt: string;
  updatedAt?: string;
}

export interface Answer {
  text: string;
  sources: string[];
  retrievalK?: number;
  embedMs?: number;
  searchMs?: number;
  llmMs?: number;
}

export interface QueryHistoryItem {
  question: string;
  answer: string;
  sources: string[];
  timestamp: string;
  success: boolean;
}

export interface Config {
  apiKey: string;
  baseUrl?: string;
  timeout?: number;
}

// Client Class
export class RAGClient {
  private apiKey: string;
  private baseUrl: string;
  private timeout: number;
  private headers: { [key: string]: string };

  constructor(config: Config) {
    this.apiKey = config.apiKey;
    this.baseUrl = (config.baseUrl || "https://api.rag-engine.com").replace(/\/$/, "");
    this.timeout = config.timeout || 30000;
    this.headers = {
      "Authorization": `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };
  }

  /**
   * Ask a question to the RAG engine.
   * @param question - Question to ask
   * @param k - Number of chunks to retrieve (default: 5)
   * @param documentId - Optional document ID for chat mode
   * @param expandQuery - Use query expansion (default: false)
   * @returns Answer with text and sources
   * 
   * طرح سؤال على محرك RAG
   */
  async ask(
    question: string,
    { k = 5, documentId, expandQuery = false }:
      {
        question: string;
        k?: number;
        documentId?: string;
        expandQuery?: boolean;
      }
  ): Promise<Answer> {
    const response = await this.fetch("/api/v1/ask", {
      method: "POST",
      body: JSON.stringify({
        question,
        k,
        document_id: documentId,
        expand_query: expandQuery,
      }),
    });

    const data = await response.json();

    return {
      text: data.answer,
      sources: data.sources || [],
      retrievalK: data.retrieval_k,
      embedMs: data.embed_ms,
      searchMs: data.search_ms,
      llmMs: data.llm_ms,
    };
  }

  /**
   * Upload a document to the RAG engine.
   * @param filePath - Path or File object to upload
   * @param filename - Optional custom filename
   * @returns Document metadata
   * 
   * رفع مستند إلى محرك RAG
   */
  async uploadDocument(
    filePath: string | File,
    { filename }:
      {
        filePath: string | File;
        filename?: string;
      }
  ): Promise<Document> {
    let file: File;

    if (typeof filePath === "string") {
      // Node.js: Read file from path
      const fs = await import("fs/promises");
      const buffer = await fs.readFile(filePath as string);
      file = new File([buffer], filename || (filePath as string).split("/").pop()!);
    } else {
      // Browser: File object provided directly
      file = filePath as File;
    }

    const formData = new FormData();
    formData.append("file", file);

    const response = await this.fetch("/api/v1/documents", {
      method: "POST",
      body: formData,
      headers: {}, // Let browser set Content-Type with boundary
    });

    const data = await response.json();

    return {
      id: data.id,
      filename: data.filename,
      contentType: data.content_type,
      sizeBytes: data.size_bytes,
      status: data.status as DocumentStatus,
      createdAt: data.created_at,
      updatedAt: data.updated_at,
    };
  }

  /**
   * Search for documents.
   * @param query - Search query
   * @param k - Number of results (default: 10)
   * @param filters - Optional filters (status, type, date, size)
   * @param sortBy - Sort order (default: created)
   * @param limit - Max results (default: 20)
   * @param offset - Pagination offset (default: 0)
   * @returns List of matching documents
   * 
   * البحث عن المستندات
   */
  async searchDocuments(
    query: string,
    {
      k = 10,
      filters,
      sortBy = QuerySortBy.CREATED,
      limit = 20,
      offset = 0,
    }:
      {
        query: string;
        k?: number;
        filters?: { [key: string]: any };
        sortBy?: QuerySortBy;
        limit?: number;
        offset?: number;
      }
  ): Promise<Document[]> {
    const params = new URLSearchParams({
      query,
      k: String(k),
      sort_by: sortBy,
      limit: String(limit),
      offset: String(offset),
    });

    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        params.append(key, String(value));
      });
    }

    const response = await this.fetch(`/api/v1/documents/search?${params.toString()}`, {
      method: "GET",
    });

    const data = await response.json();

    return data.results.map((d: any) => ({
      id: d.id,
      filename: d.filename,
      contentType: d.content_type,
      sizeBytes: d.size_bytes,
      status: d.status as DocumentStatus,
      createdAt: d.created_at,
      updatedAt: d.updated_at,
    }));
  }

  /**
   * Delete a document.
   * @param documentId - ID of document to delete
   * @returns True if successful
   * 
   * حذف مستند
   */
  async deleteDocument(documentId: string): Promise<boolean> {
    const response = await this.fetch(`/api/v1/documents/${documentId}`, {
      method: "DELETE",
    });

    return response.ok;
  }

  /**
   * Get query history.
   * @param limit - Max history items (default: 50)
   * @param offset - Pagination offset (default: 0)
   * @returns List of query history items
   * 
   * الحصول على سجل الاستعلامات
   */
  async getQueryHistory(
    { limit = 50, offset = 0 }:
      {
        limit?: number;
        offset?: number;
      }
  ): Promise<QueryHistoryItem[]> {
    const params = new URLSearchParams({
      limit: String(limit),
      offset: String(offset),
    });

    const response = await this.fetch(`/api/v1/queries/history?${params.toString()}`, {
      method: "GET",
    });

    const data = await response.json();

    return data.questions.map((q: any) => ({
      question: q.question,
      answer: q.answer || "",
      sources: q.sources || [],
      timestamp: q.timestamp,
      success: q.success ?? true,
    }));
  }

  /**
   * Helper method to make HTTP requests.
   * @param endpoint - API endpoint
   * @param options - Request options
   * @returns Response object
   */
  private async fetch(
    endpoint: string,
    options: {
      method?: string;
      body?: BodyInit | null;
      headers?: { [key: string]: string };
    }
  ): Promise<Response> {
    const url = `${this.baseUrl}${endpoint}`;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        ...options,
        headers: { ...this.headers, ...options.headers },
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return response;
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }
}

// React Hook
export function useRAGClient(config: Config) {
  const client = React.useMemo(() => new RAGClient(config), [config.apiKey]);

  return client;
}

// Example Usage
// Node.js
async function exampleNode() {
  const client = new RAGClient({
    apiKey: "sk_your_api_key_here",
    baseUrl: "http://localhost:8000",
  });

  // Ask a question
  const answer = await client.ask("What is RAG?", { k: 5 });
  console.log("Answer:", answer.text);
  console.log("Sources:", answer.sources);

  // Upload a document
  const doc = await client.uploadDocument("./test.pdf");
  console.log("Document ID:", doc.id);

  // Search documents
  const results = await client.searchDocuments("vector search");
  console.log("Results:", results.length);

  // Get query history
  const history = await client.getQueryHistory();
  console.log("History:", history.length);
}

// React Example
function RAGComponent() {
  const client = useRAGClient({
    apiKey: "sk_your_api_key_here",
  });

  const [question, setQuestion] = React.useState("");
  const [answer, setAnswer] = React.useState<Answer | null>(null);
  const [loading, setLoading] = React.useState(false);

  const handleAsk = async () => {
    setLoading(true);
    try {
      const result = await client.ask(question, { k: 5 });
      setAnswer(result);
    } catch (error) {
      console.error("Error:", error);
    } finally {
      setLoading(false);
    }
  };

  return (
    <div>
      <h1>RAG Chat</h1>
      <input
        type="text"
        value={question}
        onChange={(e) => setQuestion(e.target.value)}
        placeholder="Ask a question..."
      />
      <button onClick={handleAsk} disabled={loading}>
        {loading ? "Loading..." : "Ask"}
      </button>
      {answer && (
        <div>
          <h2>Answer</h2>
          <p>{answer.text}</p>
          <h3>Sources</h3>
          <ul>
            {answer.sources.map((source) => (
              <li key={source}>{source}</li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
}

export default RAGClient;
